nanomq/apps/broker.c

```diff
@@ -474,16 +476,146 @@ broker(const char *url)
 #endif
 }
 
-int
-broker_start(int argc, char **argv)
+void *
+print_usage(void)
 {
+	fprintf(stderr, USAGE);
+}
+
+int status_check(pid_t *pid)
+{
+	char* data = NULL;
+	size_t size = 0;
+
 	int rc;
-	if (argc != 1) {
-		fprintf(stderr, "Usage: broker start <url>\n");
+	if ((rc = nng_file_get(PID_PATH_NAME, (void*)&data, &size)) != 0) {
+		debug_msg(".pid file does not existed or cannot be read");
+		return 1;
+
+	} else {
+		if ((data) != NULL ) {
+			sscanf(data, "%lu", pid);
+			debug_msg("pid read, [%lu]", *pid);
+
+			if ((kill(*pid, 0)) == 0) {
+				debug_msg("there is a running NanoMQ instance has pid [%lu]", *pid);
+				return 0;
+			}
+		}
+		if (!nng_file_delete(PID_PATH_NAME)) {
+			debug_msg(".pid file successfully deleted");
+			return 1;
+		}
+		debug_msg("unexpected error");
+		return -1;
+	}
+}
+
+int store_pid()
+{
+	int status;
+	char pid_c[10] = "";
+
+	sprintf(pid_c, "%ld", getpid());
+	debug_msg("%s", pid_c);
+
+	status = nng_file_put(PID_PATH_NAME, pid_c, sizeof(pid_c));
+	return status;
+	
+}
+
+int broker_start(int argc, char **argv)
+{
+	int   i, url, temp, rc, num_ctx = 0;
+	pid_t pid = 0;
+
+	if (argc < 1) {
+		print_usage();
 		exit(EXIT_FAILURE);
 	}
-	rc = broker(argv[0]);
+
+	if (!status_check(&pid)) {
+		fprintf(stderr, "One NanoMQ instance is running, a new instance won't be started until the other one is stopped.\n");
+		exit(EXIT_FAILURE);
+	}
+
+	for (i = 0; i < argc; i++, temp = 0) {
+		if (argv[i][0] == '-') {
+			if (!strcmp("-daemon", argv[i]) && process_daemonize()) {
+				fprintf(stderr, "Cannot daemonize\n");
+				print_usage();
+				exit(EXIT_FAILURE);
+			}
+			else if (!strcmp("-tq_thread", argv[i]) && isdigit(argv[++i][0]) && ((temp = atoi(argv[i])) > 0)){
+				nng_taskq_setter(temp, 0);
+			}
+			else if (!strcmp("-max_tq_thread", argv[i]) && isdigit(argv[++i][0]) && ((temp = atoi(argv[i])) > 0)){
+				nng_taskq_setter(0, temp);
+			}
+			else if (!strcmp("-parallel", argv[i]) && isdigit(argv[++i][0]) && ((temp = atoi(argv[i])) > 0))
+				num_ctx = temp;
+			else {
+				print_usage();
+				exit(EXIT_FAILURE);
+			}
+		}
+		else {
+			url= i;
```

Comment:
success in your test due to the `url` had been assigned twice. first assignment is `url=start's index`, and second assignment is `url=url's index`

Reply:
Since both `argc` and `argv` have already been decremented and incremented, argv[0] will be pointed to the first argument when i = 0. So there should be no problem here.